tests/cases/conformance/types/typeParameters/typeParameterLists/typeParameterConstModifiersReverseMappedTypes.ts

```diff
@@ -0,0 +1,41 @@
+// @strict: true
+// @noEmit: true
+
+declare function test1<const T>(obj: {
+  [K in keyof T]: T[K];
+}): [T, typeof obj];
+
+const result1 = test1({
+  prop: "foo",
+  nested: {
+    nestedProp: "bar",
+  },
+});
+
+declare function test2<const T>(obj: {
+  readonly [K in keyof T]: T[K];
+}): [T, typeof obj];
+
+const result2 = test2({
+  prop: "foo",
+  nested: {
+    nestedProp: "bar",
+  },
+});
+
+declare function test3<const T>(obj: {
+  -readonly [K in keyof T]: T[K];
+}): [T, typeof obj];
+
+const result3 = test3({
+  prop: "foo",
+  nested: {
+    nestedProp: "bar",
+  },
+});
+
+declare function test4<const T extends readonly unknown[]>(arr: {
+  [K in keyof T]: T[K];
+}): T;
+
+const result4 = test4(["1", 2]);
```

Comment:
This is an especially important part of the fix. Currently `T` is inferred as `(2 | "1")[]` but it really should be `readonly ["1", 2]`